File: javaparser/src/main/java/osmparser/RawModel.java
package osmparser;

import java.util.*;

/*
    Plain data holders filled during Phase 1 (pure XML reading).
    No JTS geometry is touched here on purpose - this lets us time
    "just parsing the XML" separately from "building geometry".
*/
final class RawModel
{
    static final class RawNode
    {
        final long id;
        final double lat;
        final double lon;
        final Map<String, String> tags;

        RawNode(long id, double lat, double lon, Map<String, String> tags)
        {
            this.id = id;
            this.lat = lat;
            this.lon = lon;
            this.tags = tags;
        }
    }

    static final class RawWay
    {
        final long id;
        final Map<String, String> tags;
        final List<Long> nodeRefs;

        RawWay(long id, Map<String, String> tags, List<Long> nodeRefs)
        {
            this.id = id;
            this.tags = tags;
            this.nodeRefs = nodeRefs;
        }
    }

    static final class RawRelationMember
    {
        final String type;
        final String role;
        final long ref;

        RawRelationMember(String type, String role, long ref)
        {
            this.type = type;
            this.role = role;
            this.ref = ref;
        }
    }

    static final class RawRelation
    {
        final long id;
        final Map<String, String> tags;
        final List<RawRelationMember> members;

        RawRelation(long id, Map<String, String> tags, List<RawRelationMember> members)
        {
            this.id = id;
            this.tags = tags;
            this.members = members;
        }
    }

    final List<RawNode> nodes = new ArrayList<>();
    final List<RawWay> ways = new ArrayList<>();
    final List<RawRelation> relations = new ArrayList<>();
}

File: rustparser/src/raw_model.rs
use std::collections::HashMap;

pub struct RawNode {
    pub id: i64,
    pub lat: f64,
    pub lon: f64,
    pub tags: HashMap<String, String>,
}

pub struct RawWay {
    pub id: i64,
    pub tags: HashMap<String, String>,
    pub node_refs: Vec<i64>,
}

pub struct RawMember {
    pub member_type: String,
    pub role: String,
    pub member_ref: i64,
}

pub struct RawRelation {
    pub id: i64,
    pub tags: HashMap<String, String>,
    pub members: Vec<RawMember>,
}

#[derive(Default)]
pub struct RawModel {
    pub nodes: Vec<RawNode>,
    pub ways: Vec<RawWay>,
    pub relations: Vec<RawRelation>,
}
